OneCommunityGlobal/HGNRest
# Abdel streamlined data input for new members

# Description

#### $\textcolor{red}{\textsf{(PRIORITY HIGH) Jae:}}$: Create a way for easy/nearly-automated data input for creating a new member 
- I’d like to be able to have a way for people to input their own information for initial setup in the software. Right now I set them up manually and it takes me a fair amount of time. It would be better to be able to send them a 1-time link that would have them set up the information for themselves. 
- Link needs to be 1-time so that people can’t just randomly create users
- Link should be generated by me just inputting their email somewhere like the top of the User Management page
- The parts that will need to be set up by me after they set themselves up are their Team and Projects/Tasks, so they shouldn’t have access to choose these. The rest though they should set up themselves. 
- I’d like to get an email each time a person sets themselves up, so I can easily check the details without logging into the software. If errors were made, I could also ask them to correct them by just forwarding the email with the corrections. 

## Related PRS:
This backend PR is related to the [#1073](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/1073) frontend PR.

## Main changes explained:
- Updated routes.js and middleware.js.
- Added profileInitialSetupController.js, profileInitialSetupRouter.js and profileInitialSetupToken.js.

## How to test:
- check into the current branch.
- Set the environment variables (.env file) as follow:
- First you need to set up your own test email id to send the mail notification.

- To send email using Gmail API using OAuth2 and nodemailer, kindly follow the below video and setup up a test mail id.
https://www.youtube.com/watch?v=-rcRf7yswfM&t=10s&ab_channel=MafiaCodes

- Fill below variable values from the above setup in the .env file of the backend.
```
REACT_APP_EMAIL =
REACT_APP_EMAIL_CLIENT_ID =
REACT_APP_EMAIL_CLIENT_SECRET =
REACT_APP_EMAIL_REFRESH_TOKEN =
REACT_APP_EMAIL_CLIENT_REDIRECT_URI = https://developers.google.com/oauthplayground
sendEmail = TRUE
```
- Fill below variable value with the email address you wish to use for receiving notifications
```
MANAGER_EMAIL=<add your own email>
```

- Get a free API key from https://opencagedata.com/pricing
- Fill below variable value with the API key
```
TIMEZONE_PREMIUM_KEY=<add API key here>
```
- npm run build
- npm start
